File: Integrative Gene Expression Modeling from DNA Sequence Features/scripts/gene_expression_prediction.R
# Gene Expression Prediction from DNA Sequences
# =============================================
# This script predicts gene expression levels from DNA sequences using various
# machine learning approaches in R.

# Load required libraries
library(readr)
library(dplyr)
library(ggplot2)
library(caret)
library(randomForest)
library(xgboost)
library(e1071)
library(pROC)
library(doParallel)

# Set random seed for reproducibility
set.seed(42)

# ============================================================================
# 1. DATA LOADING AND PREPROCESSING
# ============================================================================

cat(paste0(rep("=", 80), collapse = ""), "\n")
cat("Gene Expression Prediction from DNA Sequences\n")
cat(paste0(rep("=", 80), collapse = ""), "\n\n")

# Load data
cat("1. Loading data...\n")
# Determine the correct data path based on where the script is run from
# If running from project root: use "data/genomics_data.csv"
# If running from scripts directory: use "../data/genomics_data.csv"
data_path <- file.path("data", "genomics_data.csv")
if (!file.exists(data_path)) {
  data_path <- file.path("..", "data", "genomics_data.csv")
}

if (!file.exists(data_path)) {
  stop("Error: Could not find genomics_data.csv. Please ensure you're running from the project root directory.")
}

df <- read_csv(data_path, show_col_types = FALSE)

cat(sprintf("Dataset shape: %d rows, %d columns\n", nrow(df), ncol(df)))
cat(sprintf("Label distribution:\n"))
print(table(df$Labels))

# ============================================================================
# 2. FEATURE ENGINEERING
# ============================================================================

cat("\n2. Encoding DNA sequences...\n")

# K-mer encoding function
kmer_encoding <- function(sequences, k = 3) {
  # Extract k-mers from each sequence
  kmers_list <- lapply(sequences, function(seq) {
    seq_length <- nchar(seq)
    if (seq_length < k) {
      return(character(0))
    }
    kmers <- character(seq_length - k + 1)
    for (i in 1:(seq_length - k + 1)) {
      kmers[i] <- substr(seq, i, i + k - 1)
    }
    return(kmers)
  })
  
  # Create a list of all unique k-mers
  all_kmers <- unique(unlist(kmers_list))
  
  # Create frequency matrix
  n_sequences <- length(sequences)
  n_kmers <- length(all_kmers)
  freq_matrix <- matrix(0, nrow = n_sequences, ncol = n_kmers)
  colnames(freq_matrix) <- all_kmers
  
  # Count k-mer frequencies
  for (i in 1:n_sequences) {
    kmers <- kmers_list[[i]]
    if (length(kmers) > 0) {
      kmers_table <- table(kmers)
      for (kmer in names(kmers_table)) {
        if (kmer %in% all_kmers) {
          freq_matrix[i, kmer] <- kmers_table[kmer]
        }
      }
      # Normalize by sequence length
      freq_matrix[i, ] <- freq_matrix[i, ] / length(kmers)
    }
  }
  
  return(list(matrix = freq_matrix, kmers = all_kmers))
}

# Nucleotide composition function
nucleotide_composition <- function(sequences) {
  nucleotides <- c("A", "T", "G", "C")
  n_sequences <- length(sequences)
  n_features <- 4 + 1  # 4 nucleotides + GC content
  composition_matrix <- matrix(0, nrow = n_sequences, ncol = n_features)
  colnames(composition_matrix) <- c(nucleotides, "GC_content")
  
  for (i in 1:n_sequences) {
    seq <- toupper(sequences[i])
    seq_chars <- strsplit(seq, "")[[1]]
    seq_length <- length(seq_chars)
    
    # Nucleotide frequencies
    for (j in 1:4) {
      composition_matrix[i, j] <- sum(seq_chars == nucleotides[j]) / seq_length
    }
    
    # GC content
    composition_matrix[i, 5] <- (composition_matrix[i, 3] + composition_matrix[i, 4])
  }
  
  return(composition_matrix)
}

# Encode sequences
sequences <- df$Sequences
labels <- as.factor(df$Labels)

cat("Encoding sequences using k-mer approach (k=3)...\n")
kmer_result <- kmer_encoding(sequences, k = 3)
X <- kmer_result$matrix
cat(sprintf("Encoded feature shape: %d x %d\n", nrow(X), ncol(X)))

# ============================================================================
# 3. DATA SPLITTING
# ============================================================================

cat("\n3. Splitting data into train and test sets...\n")

# Create data frame for modeling
model_data <- as.data.frame(X)
model_data$Label <- labels

# Split data
train_indices <- createDataPartition(model_data$Label, p = 0.8, list = FALSE)
train_data <- model_data[train_indices, ]
test_data <- model_data[-train_indices, ]

X_train <- train_data[, !colnames(train_data) %in% "Label"]
y_train <- train_data$Label
X_test <- test_data[, !colnames(test_data) %in% "Label"]
y_test <- test_data$Label

cat(sprintf("Training set: %d samples\n", nrow(train_data)))
cat(sprintf("Test set: %d samples\n", nrow(test_data)))
cat(sprintf("Training label distribution: %s\n", paste(table(y_train), collapse = ", ")))
cat(sprintf("Test label distribution: %s\n", paste(table(y_test), collapse = ", ")))

# ============================================================================
# 4. MODEL TRAINING AND EVALUATION
# ============================================================================

cat("\n4. Training models...\n")

# Function to evaluate model
evaluate_model <- function(model, X_test, y_test, model_name, test_data_frame = NULL) {
  # Initialize variables
  y_pred <- NULL
  y_pred_proba <- NULL
  
  # Predictions
  if (class(model)[1] == "xgb.Booster") {
    # XGBoost requires matrix input
    test_matrix <- as.matrix(X_test)
    y_pred_proba <- predict(model, test_matrix)
    y_pred <- ifelse(y_pred_proba > 0.5, 1, 0)
    y_pred <- factor(y_pred, levels = levels(y_test))
  } else if (class(model)[1] == "randomForest") {
    # Random Forest
    y_pred <- predict(model, test_data_frame)
    prob_matrix <- predict(model, test_data_frame, type = "prob")
    if (ncol(prob_matrix) >= 2) {
      y_pred_proba <- prob_matrix[, 2]
    } else {
      y_pred_proba <- as.numeric(y_pred) - 1
    }
  } else if (class(model)[1] == "svm") {
    # SVM
    y_pred <- predict(model, test_data_frame)
    # Try to get probabilities from SVM
    tryCatch({
      svm_pred_prob <- predict(model, test_data_frame, probability = TRUE)
      prob_attr <- attr(svm_pred_prob, "probabilities")
      if (!is.null(prob_attr) && ncol(prob_attr) >= 2) {
        # Get probability for class "1" (high expression)
        if ("1" %in% colnames(prob_attr)) {
          y_pred_proba <- prob_attr[, "1"]
        } else {
          y_pred_proba <- prob_attr[, 2]
        }
      } else {
        # Fallback: convert predictions to numeric probabilities
        y_pred_proba <- as.numeric(y_pred) - 1
      }
    }, error = function(e) {
      # If probability extraction fails, use predictions as probabilities
      y_pred_proba <<- as.numeric(y_pred) - 1
    })
    
    # Ensure y_pred_proba is set
    if (is.null(y_pred_proba)) {
      y_pred_proba <- as.numeric(y_pred) - 1
    }
  } else {
    # Default: try to get predictions
    if (!is.null(test_data_frame)) {
      y_pred <- predict(model, test_data_frame)
      # Try to get probabilities
      tryCatch({
        prob_result <- predict(model, test_data_frame, type = "prob")
        if (is.matrix(prob_result) && ncol(prob_result) >= 2) {
          y_pred_proba <- prob_result[, 2]
        } else {
          y_pred_proba <- as.numeric(y_pred) - 1
        }
      }, error = function(e) {
        y_pred_proba <<- as.numeric(y_pred) - 1
      })
    } else {
      y_pred <- predict(model, X_test)
      y_pred_proba <- as.numeric(y_pred) - 1
    }
  }
  
  # Ensure both variables are set
  if (is.null(y_pred) || is.null(y_pred_proba)) {
    stop(sprintf("Error: Could not generate predictions for %s", model_name))
  }
  
  # Metrics
  accuracy <- mean(y_pred == y_test)
  cm <- confusionMatrix(y_pred, y_test)
  
  # ROC AUC
  roc_auc <- NA
  tryCatch({
    roc_obj <- roc(as.numeric(y_test) - 1, y_pred_proba, quiet = TRUE)
    roc_auc <- as.numeric(auc(roc_obj))
  }, error = function(e) {
    cat(sprintf("Warning: Could not calculate ROC-AUC for %s\n", model_name))
  })
  
  return(list(
    accuracy = accuracy,
    confusion_matrix = cm,
    y_pred = y_pred,
    y_pred_proba = y_pred_proba,
    roc_auc = roc_auc
  ))
}

# Train models
results <- list()

# Random Forest
cat("\n   Training Random Forest...\n")
rf_model <- randomForest(
  Label ~ .,
  data = train_data,
  ntree = 100,
  mtry = sqrt(ncol(X_train)),
  importance = TRUE
)
rf_results <- evaluate_model(rf_model, X_test, y_test, "Random Forest", test_data)
results[["Random Forest"]] <- list(model = rf_model, metrics = rf_results)
cat(sprintf("   Accuracy: %.4f\n", rf_results$accuracy))
if (!is.na(rf_results$roc_auc)) {
  cat(sprintf("   ROC-AUC: %.4f\n", rf_results$roc_auc))
}

# XGBoost
cat("\n   Training XGBoost...\n")
train_matrix <- as.matrix(X_train)
test_matrix <- as.matrix(X_test)
dtrain <- xgb.DMatrix(data = train_matrix, label = as.numeric(y_train) - 1)
dtest <- xgb.DMatrix(data = test_matrix, label = as.numeric(y_test) - 1)

xgb_model <- xgb.train(
  data = dtrain,
  nrounds = 100,
  objective = "binary:logistic",
  eval_metric = "logloss",
  max_depth = 6,
  eta = 0.3,
  verbose = 0
)
xgb_results <- evaluate_model(xgb_model, X_test, y_test, "XGBoost")
results[["XGBoost"]] <- list(model = xgb_model, metrics = xgb_results)
cat(sprintf("   Accuracy: %.4f\n", xgb_results$accuracy))
if (!is.na(xgb_results$roc_auc)) {
  cat(sprintf("   ROC-AUC: %.4f\n", xgb_results$roc_auc))
}

# SVM
cat("\n   Training SVM...\n")
svm_model <- svm(
  Label ~ .,
  data = train_data,
  kernel = "radial",
  probability = TRUE
)
svm_results <- evaluate_model(svm_model, X_test, y_test, "SVM", test_data)
results[["SVM"]] <- list(model = svm_model, metrics = svm_results)
cat(sprintf("   Accuracy: %.4f\n", svm_results$accuracy))
if (!is.na(svm_results$roc_auc)) {
  cat(sprintf("   ROC-AUC: %.4f\n", svm_results$roc_auc))
}

# ============================================================================
# 5. RESULTS AND VISUALIZATION
# ============================================================================

cat("\n5. Generating plots...\n")

# Create results directory if it doesn't exist
# Try project root first, then scripts directory parent
results_dir <- ifelse(dir.exists("results"), "results", file.path("..", "results"))
if (!dir.exists(results_dir)) {
  dir.create(results_dir, recursive = TRUE)
}

# Accuracy comparison
accuracies <- sapply(results, function(x) x$metrics$accuracy)
roc_aucs <- sapply(results, function(x) {
  if (!is.na(x$metrics$roc_auc)) {
    return(x$metrics$roc_auc)
  } else {
    return(NA)
  }
})

# Plot accuracy comparison
png(file.path(results_dir, "model_comparison_R.png"), width = 1200, height = 600)
par(mfrow = c(1, 2))

barplot(accuracies, 
        names.arg = names(accuracies),
        col = c("#3498db", "#e74c3c", "#2ecc71"),
        main = "Model Accuracy Comparison",
        ylab = "Accuracy",
        ylim = c(0, 1),
        las = 2)
text(x = 1:length(accuracies), 
     y = accuracies + 0.02,
     labels = sprintf("%.3f", accuracies),
     cex = 0.8)

# Plot ROC-AUC comparison (if available)
if (any(!is.na(roc_aucs))) {
  valid_aucs <- roc_aucs[!is.na(roc_aucs)]
  valid_names <- names(roc_aucs)[!is.na(roc_aucs)]
  barplot(valid_aucs,
          names.arg = valid_names,
          col = c("#3498db", "#e74c3c", "#2ecc71")[1:length(valid_aucs)],
          main = "Model ROC-AUC Comparison",
          ylab = "ROC-AUC",
          ylim = c(0, 1),
          las = 2)
  text(x = 1:length(valid_aucs),
       y = valid_aucs + 0.02,
       labels = sprintf("%.3f", valid_aucs),
       cex = 0.8)
}

dev.off()

# ROC curves
png(file.path(results_dir, "roc_curves_R.png"), width = 1000, height = 800)
plot(1, type = "n", 
     xlim = c(0, 1), ylim = c(0, 1),
     xlab = "False Positive Rate",
     ylab = "True Positive Rate",
     main = "ROC Curves")
abline(a = 0, b = 1, lty = 2, col = "gray")

colors <- c("#3498db", "#e74c3c", "#2ecc71")
i <- 1
for (model_name in names(results)) {
  metrics <- results[[model_name]]$metrics
  if (!is.na(metrics$roc_auc)) {
    roc_obj <- roc(as.numeric(y_test) - 1, metrics$y_pred_proba, quiet = TRUE)
    lines(roc_obj, col = colors[i], lwd = 2)
    i <- i + 1
  }
}

legend("bottomright",
       legend = paste(names(results), 
                     sprintf("(AUC = %.3f)", roc_aucs)),
       col = colors[1:length(results)],
       lwd = 2)
dev.off()

# ============================================================================
# 6. SAVE MODELS
# ============================================================================

cat("\n6. Saving models...\n")
# Try project root first, then scripts directory parent
models_dir <- ifelse(dir.exists("models"), "models", file.path("..", "models"))
if (!dir.exists(models_dir)) {
  dir.create(models_dir, recursive = TRUE)
}

# Find best model
best_model_name <- names(results)[which.max(accuracies)]
best_model <- results[[best_model_name]]$model

# Save best model
if (best_model_name == "XGBoost") {
  xgb.save(best_model, file.path(models_dir, "best_model_xgboost_R.model"))
} else {
  saveRDS(best_model, file.path(models_dir, "best_model_R.rds"))
}

cat(sprintf("Best model (%s) saved to %s\n", 
            best_model_name, 
            file.path(models_dir, "best_model_R.rds")))

# ============================================================================
# 7. PRINT DETAILED RESULTS
# ============================================================================

cat("\n7. Detailed Results:\n")
cat(strrep("=", 80) %+% "\n")

for (model_name in names(results)) {
  cat(sprintf("\n%s:\n", model_name))
  cat(sprintf("  Accuracy: %.4f\n", results[[model_name]]$metrics$accuracy))
  if (!is.na(results[[model_name]]$metrics$roc_auc)) {
    cat(sprintf("  ROC-AUC: %.4f\n", results[[model_name]]$metrics$roc_auc))
  }
  cat("\n  Confusion Matrix:\n")
  print(results[[model_name]]$metrics$confusion_matrix)
}

cat("\n", paste0(rep("=", 80), collapse = ""), "\n")
cat("Pipeline completed successfully!\n")
cat(paste0(rep("=", 80), collapse = ""), "\n")

File: Integrative Gene Expression Modeling from DNA Sequence Features/scripts/gene_expression_prediction.py
"""
Gene Expression Prediction from DNA Sequences
=============================================
This script predicts gene expression levels from DNA sequences using various
machine learning approaches including traditional ML and deep learning models.
"""

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, roc_auc_score, roc_curve
from sklearn.feature_extraction.text import CountVectorizer
import xgboost as xgb
import lightgbm as lgb
import joblib
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path
import warnings
warnings.filterwarnings('ignore')

# Set random seed for reproducibility
np.random.seed(42)

class DNASequenceEncoder:
    """Encode DNA sequences into numerical features."""
    
    def __init__(self, method='kmer'):
        """
        Initialize encoder.
        
        Parameters:
        -----------
        method : str
            Encoding method: 'kmer', 'onehot', or 'nucleotide_composition'
        """
        self.method = method
        self.vectorizer = None
        self.kmer_size = 3
        
    def kmer_encoding(self, sequences, k=3):
        """
        Convert DNA sequences to k-mer frequency features.
        
        Parameters:
        -----------
        sequences : list
            List of DNA sequences
        k : int
            k-mer size (default: 3 for trinucleotides)
            
        Returns:
        --------
        array : numpy array of k-mer frequencies
        """
        kmers = []
        for seq in sequences:
            seq_kmers = [seq[i:i+k] for i in range(len(seq) - k + 1)]
            kmers.append(' '.join(seq_kmers))
        
        if self.vectorizer is None:
            self.vectorizer = CountVectorizer(analyzer='word', token_pattern=r'\S+')
            X = self.vectorizer.fit_transform(kmers).toarray()
        else:
            X = self.vectorizer.transform(kmers).toarray()
        
        return X
    
    def nucleotide_composition(self, sequences):
        """
        Calculate nucleotide composition features.
        
        Parameters:
        -----------
        sequences : list
            List of DNA sequences
            
        Returns:
        --------
        array : numpy array of nucleotide composition features
        """
        features = []
        nucleotides = ['A', 'T', 'G', 'C']
        
        for seq in sequences:
            seq = seq.upper()
            seq_features = []
            
            # Nucleotide frequencies
            for nuc in nucleotides:
                seq_features.append(seq.count(nuc) / len(seq))
            
            # GC content
            gc_content = (seq.count('G') + seq.count('C')) / len(seq)
            seq_features.append(gc_content)
            
            # Dinucleotide frequencies
            for nuc1 in nucleotides:
                for nuc2 in nucleotides:
                    dinuc = nuc1 + nuc2
                    seq_features.append(seq.count(dinuc) / (len(seq) - 1) if len(seq) > 1 else 0)
            
            features.append(seq_features)
        
        return np.array(features)
    
    def onehot_encoding(self, sequences, max_length=50):
        """
        One-hot encode DNA sequences.
        
        Parameters:
        -----------
        sequences : list
            List of DNA sequences
        max_length : int
            Maximum sequence length (default: 50)
            
        Returns:
        --------
        array : numpy array of one-hot encoded sequences
        """
        nuc_to_int = {'A': 0, 'T': 1, 'G': 2, 'C': 3, 'N': 4}
        onehot = np.zeros((len(sequences), max_length, 4))
        
        for i, seq in enumerate(sequences):
            seq = seq.upper()
            for j, nuc in enumerate(seq[:max_length]):
                if nuc in nuc_to_int and nuc_to_int[nuc] < 4:
                    onehot[i, j, nuc_to_int[nuc]] = 1
        
        # Flatten for traditional ML models
        return onehot.reshape(len(sequences), -1)
    
    def encode(self, sequences, fit=True):
        """
        Encode sequences using the specified method.
        
        Parameters:
        -----------
        sequences : list
            List of DNA sequences
        fit : bool
            Whether to fit the encoder (for training data)
            
        Returns:
        --------
        array : Encoded sequences
        """
        if self.method == 'kmer':
            return self.kmer_encoding(sequences, k=self.kmer_size)
        elif self.method == 'nucleotide_composition':
            return self.nucleotide_composition(sequences)
        elif self.method == 'onehot':
            return self.onehot_encoding(sequences)
        else:
            raise ValueError(f"Unknown encoding method: {self.method}")

def load_data(data_path):
    """
    Load and preprocess the genomics data.
    
    Parameters:
    -----------
    data_path : str
        Path to the CSV file
        
    Returns:
    --------
    X : numpy array
        Features (DNA sequences)
    y : numpy array
        Labels (gene expression levels)
    """
    df = pd.read_csv(data_path)
    print(f"Dataset shape: {df.shape}")
    print(f"Label distribution:\n{df['Labels'].value_counts()}")
    
    sequences = df['Sequences'].tolist()
    labels = df['Labels'].values
    
    return sequences, labels

def evaluate_model(model, X_test, y_test, model_name):
    """
    Evaluate a trained model.
    
    Parameters:
    -----------
    model : sklearn model
        Trained model
    X_test : numpy array
        Test features
    y_test : numpy array
        Test labels
    model_name : str
        Name of the model
        
    Returns:
    --------
    dict : Dictionary containing evaluation metrics
    """
    y_pred = model.predict(X_test)
    y_pred_proba = model.predict_proba(X_test)[:, 1] if hasattr(model, 'predict_proba') else None
    
    accuracy = accuracy_score(y_test, y_pred)
    
    metrics = {
        'model_name': model_name,
        'accuracy': accuracy,
        'classification_report': classification_report(y_test, y_pred),
        'confusion_matrix': confusion_matrix(y_test, y_pred),
        'y_pred': y_pred,
        'y_pred_proba': y_pred_proba
    }
    
    if y_pred_proba is not None:
        metrics['roc_auc'] = roc_auc_score(y_test, y_pred_proba)
        metrics['roc_curve'] = roc_curve(y_test, y_pred_proba)
    
    return metrics

def plot_results(metrics_dict, save_path=None):
    """
    Plot evaluation results.
    
    Parameters:
    -----------
    metrics_dict : dict
        Dictionary containing metrics for different models
    save_path : str
        Path to save the plot
    """
    fig, axes = plt.subplots(2, 2, figsize=(15, 12))
    
    # Accuracy comparison
    model_names = list(metrics_dict.keys())
    accuracies = [metrics_dict[m]['accuracy'] for m in model_names]
    
    axes[0, 0].bar(model_names, accuracies, color=['#3498db', '#e74c3c', '#2ecc71', '#f39c12', '#9b59b6'])
    axes[0, 0].set_title('Model Accuracy Comparison', fontsize=14, fontweight='bold')
    axes[0, 0].set_ylabel('Accuracy')
    axes[0, 0].set_ylim([0, 1])
    axes[0, 0].grid(axis='y', alpha=0.3)
    
    for i, (name, acc) in enumerate(zip(model_names, accuracies)):
        axes[0, 0].text(i, acc + 0.01, f'{acc:.3f}', ha='center', fontweight='bold')
    
    # ROC Curves
    for model_name, metrics in metrics_dict.items():
        if 'roc_curve' in metrics:
            fpr, tpr, _ = metrics['roc_curve']
            auc = metrics.get('roc_auc', 0)
            axes[0, 1].plot(fpr, tpr, label=f'{model_name} (AUC = {auc:.3f})', linewidth=2)
    
    axes[0, 1].plot([0, 1], [0, 1], 'k--', label='Random')
    axes[0, 1].set_xlabel('False Positive Rate')
    axes[0, 1].set_ylabel('True Positive Rate')
    axes[0, 1].set_title('ROC Curves', fontsize=14, fontweight='bold')
    axes[0, 1].legend()
    axes[0, 1].grid(alpha=0.3)
    
    # Confusion Matrix for best model
    best_model = max(metrics_dict.items(), key=lambda x: x[1]['accuracy'])
    cm = best_model[1]['confusion_matrix']
    sns.heatmap(cm, annot=True, fmt='d', cmap='Blues', ax=axes[1, 0])
    axes[1, 0].set_title(f'Confusion Matrix - {best_model[0]}', fontsize=14, fontweight='bold')
    axes[1, 0].set_ylabel('True Label')
    axes[1, 0].set_xlabel('Predicted Label')
    
    # Feature importance (for tree-based models)
    if hasattr(best_model, 'feature_importances_'):
        # This would require the model object, which we don't store in metrics_dict
        # So we'll skip this for now
        axes[1, 1].text(0.5, 0.5, 'Feature Importance\n(available for tree-based models)', 
                       ha='center', va='center', fontsize=12)
        axes[1, 1].set_title('Feature Importance', fontsize=14, fontweight='bold')
    
    plt.tight_layout()
    if save_path:
        plt.savefig(save_path, dpi=300, bbox_inches='tight')
    plt.close()

def main():
    """Main function to run the gene expression prediction pipeline."""
    
    # Setup paths
    base_path = Path(__file__).parent.parent
    data_path = base_path / 'data' / 'genomics_data.csv'
    results_path = base_path / 'results'
    models_path = base_path / 'models'
    
    results_path.mkdir(exist_ok=True)
    models_path.mkdir(exist_ok=True)
    
    print("=" * 80)
    print("Gene Expression Prediction from DNA Sequences")
    print("=" * 80)
    
    # Load data
    print("\n1. Loading data...")
    sequences, labels = load_data(data_path)
    
    # Encode sequences
    print("\n2. Encoding DNA sequences...")
    encoder = DNASequenceEncoder(method='kmer')
    X = encoder.encode(sequences, fit=True)
    print(f"Encoded feature shape: {X.shape}")
    
    # Split data
    print("\n3. Splitting data into train and test sets...")
    X_train, X_test, y_train, y_test = train_test_split(
        X, labels, test_size=0.2, random_state=42, stratify=labels
    )
    print(f"Training set: {X_train.shape[0]} samples")
    print(f"Test set: {X_test.shape[0]} samples")
    
    # Train multiple models
    print("\n4. Training models...")
    models = {
        'Random Forest': RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1),
        'Gradient Boosting': GradientBoostingClassifier(n_estimators=100, random_state=42),
        'XGBoost': xgb.XGBClassifier(n_estimators=100, random_state=42, eval_metric='logloss'),
        'LightGBM': lgb.LGBMClassifier(n_estimators=100, random_state=42, verbose=-1),
        'SVM': SVC(probability=True, random_state=42, kernel='rbf')
    }
    
    metrics_dict = {}
    trained_models = {}
    
    for name, model in models.items():
        print(f"\n   Training {name}...")
        model.fit(X_train, y_train)
        trained_models[name] = model
        
        # Evaluate
        metrics = evaluate_model(model, X_test, y_test, name)
        metrics_dict[name] = metrics
        
        print(f"   Accuracy: {metrics['accuracy']:.4f}")
        if 'roc_auc' in metrics:
            print(f"   ROC-AUC: {metrics['roc_auc']:.4f}")
    
    # Save best model
    best_model_name = max(metrics_dict.items(), key=lambda x: x[1]['accuracy'])[0]
    best_model = trained_models[best_model_name]
    model_file = models_path / f'best_model_{best_model_name.lower().replace(" ", "_")}.pkl'
    joblib.dump(best_model, model_file)
    joblib.dump(encoder, models_path / 'encoder.pkl')
    print(f"\n5. Best model ({best_model_name}) saved to {model_file}")
    
    # Plot results
    print("\n6. Generating plots...")
    plot_results(metrics_dict, save_path=results_path / 'model_comparison.png')
    print(f"   Plots saved to {results_path / 'model_comparison.png'}")
    
    # Print detailed results
    print("\n7. Detailed Results:")
    print("=" * 80)
    for name, metrics in metrics_dict.items():
        print(f"\n{name}:")
        print(f"  Accuracy: {metrics['accuracy']:.4f}")
        if 'roc_auc' in metrics:
            print(f"  ROC-AUC: {metrics['roc_auc']:.4f}")
        print(f"\n  Classification Report:\n{metrics['classification_report']}")
    
    print("\n" + "=" * 80)
    print("Pipeline completed successfully!")
    print("=" * 80)

if __name__ == "__main__":
    main()

